link-parent/link-core/src/main/java/org/interledger/link/LinkFactoryProvider.java

```diff
@@ -0,0 +1,42 @@
+package org.interledger.link;
+
+import org.interledger.link.exceptions.LinkException;
+
+import com.google.common.collect.Maps;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A provider of {@link LinkFactory} scoped by type.
+ */
+public class LinkFactoryProvider {
+
+  // At runtime, there will potentially be _many_ factories depending on the LinkType (e.g., a BTP Factory, a
+  // LoopbackFactory, etc).
+  private final Map<LinkType, LinkFactory> linkFactories;
+
+  public LinkFactoryProvider() {
+    this(Maps.newConcurrentMap());
+  }
+
+  public LinkFactoryProvider(final Map<LinkType, LinkFactory> linkFactories) {
+    this.linkFactories = Objects.requireNonNull(linkFactories);
+  }
+
+  public LinkFactory getLinkFactory(final LinkType linkType) {
+    Objects.requireNonNull(linkType, "linkType must not be null");
+    return Optional.ofNullable(this.linkFactories.get(linkType))
+        .orElseThrow(() -> new LinkException(
+            String.format("No registered LinkFactory linkType=%s", linkType), LinkId.of("n/a"))
+        );
+  }
+
+  public LinkFactory registerLinkFactory(final LinkType linkType, final LinkFactory linkFactory) {
+    Objects.requireNonNull(linkType, "linkType must not be null");
+    Objects.requireNonNull(linkFactory, "linkFactory must not be null");
+
+    return this.linkFactories.put(linkType, linkFactory);
```

Comment:
Simple answer is I'm not sure. In the Connector implementation at least (which is the only think using this) factories are only registered at startup time and then never change.

That said, we should probably leave this alone just in case some other use case comes along....or until we see this causing problems.